File: src/kiosk.js
import qrcode from 'qrcode-generator';
import { issueToken, signedInMember, startLogin, verifyToken } from './auth.js';
import { boundedText, cookie, escapeHTML as e, hash, HttpError, json, now, origin, randomToken } from './http.js';
import { fobClaim } from './fob-claims.js';
import { requireKioskNetwork } from './kiosk-network.js';
import { coordinated } from './membership.js';

function page(title, body, kiosk = false) {
  const content = kiosk ? body : `<main class="container membership-main"><a class="membership-brand" href="/"><img src="/assets/favicon.svg" alt="" width="48" height="48">TheLab</a><section class="card membership-message"><h1>${e(title)}</h1>${body}</section></main>`;
  return new Response(`<!doctype html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"><meta name="robots" content="noindex"><title>${e(title)} | TheLab</title><link rel="icon" href="/assets/favicon.svg"><link rel="stylesheet" href="/style.css"><link rel="stylesheet" href="/membership.css"><link rel="stylesheet" href="/kiosk.css">${kiosk ? '<script src="/kiosk.js" defer></script>' : ''}</head><body class="membership-page${kiosk ? ' kiosk-page' : ''}">${content}</body></html>`, {
    headers: { 'Content-Type': 'text/html; charset=utf-8', 'Cache-Control': 'no-store', 'Referrer-Policy': 'no-referrer',
      'X-Content-Type-Options': 'nosniff', 'Content-Security-Policy': "default-src 'none'; script-src 'self'; style-src 'self'; img-src 'self' data:; font-src 'self'; connect-src 'self'; form-action 'self'; base-uri 'none'; frame-ancestors 'none'" },
  });
}

function sameOrigin(request, env) {
  if (request.headers.get('Origin') !== origin(env)) throw new HttpError(403, 'We couldn’t process your request. Reload this page and try again.');
}

export function cleanupFobClaims(env) {
  return env.DB.prepare('DELETE FROM fob_claims WHERE expires <= ?').bind(now()).run();
}

export async function kioskPage(request, env) {
  await requireKioskNetwork(request, env);
  return page('Link your key fob', `<div class="kiosk-shell">
    <header class="kiosk-header"><div class="kiosk-brand"><img src="/assets/favicon.svg" alt="" width="48" height="48">TheLab<span>Everyone’s makerspace</span></div><span class="kiosk-label">Key fob station</span></header>
    <main id="kiosk" class="kiosk-main" data-state="ready">
      <section id="standby" class="kiosk-standby" aria-labelledby="kiosk-title">
        <div class="kiosk-reader" aria-hidden="true"><svg viewBox="0 0 120 120" fill="none" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"><rect x="24" y="39" width="56" height="40" rx="8" transform="rotate(-12 52 59)"/><circle cx="39" cy="60" r="5"/><path d="M86 43a25 25 0 0 1 0 34M96 33a39 39 0 0 1 0 54"/></svg></div>
        <p class="kiosk-eyebrow">Make yourself at home</p><h1 id="kiosk-title">Tap your fob.<br><span>Start making.</span></h1>
        <p class="kiosk-intro">Hold your key fob near the reader<br>to link it to your membership.</p>
      </section>
      <section id="claim" class="kiosk-claim" aria-labelledby="claim-title" hidden>
        <div class="kiosk-claim-copy"><p class="kiosk-eyebrow">One more step</p><h1 id="claim-title">Finish on<br>your phone.</h1><p>Scan this QR code with your phone’s camera, sign in through Discord, then tap <strong>Link fob</strong>.</p><p id="expiry" class="kiosk-expiry"></p><button id="done" type="button" class="btn btn-outline">Cancel / start over</button></div>
        <div class="kiosk-qr-card"><img id="qr" alt="Scan to link your key fob"><p>Your membership. Your key.</p></div>
      </section>
      <div class="kiosk-status"><span class="kiosk-status-dot" aria-hidden="true"></span><p id="status" role="status" aria-live="polite">Ready when you are. Just tap your fob.</p></div>
      <noscript><p class="kiosk-noscript">Enable JavaScript to use the fob reader and display QR codes.</p></noscript>
    </main>
    <footer class="kiosk-footer"><ol><li><span>01</span> Tap your fob</li><li><span>02</span> Scan the QR code</li><li><span>03</span> Link with Discord</li></ol><p>Need a hand? Ask a member.</p></footer>
    </div>`, true);
}

export async function kioskClaims(request, env) {
  await requireKioskNetwork(request, env);
  if (request.method === 'GET') {
    const params = new URL(request.url).searchParams;
    if (params.getAll('token').length !== 1) throw new HttpError(400, 'Scan your fob at the kiosk to get started.');
    const claim = await fobClaim(env, params.get('token'));
    return json({ claimed: claim.claimed_by !== null, expires: claim.expires });
  }
  sameOrigin(request, env);
  if (request.headers.get('Content-Type')?.split(';')[0] !== 'application/json') throw new HttpError(415, 'Fob not recognized. Scan again.');
  let input;
  try { input = JSON.parse(await boundedText(request, 1024)); }
  catch (error) { if (error instanceof HttpError) throw error; throw new HttpError(400, 'Fob not recognized. Scan again.'); }
  const fob = input?.fob_id;
  if (!Number.isInteger(fob) || fob < 1 || fob > 4294967295) throw new HttpError(400, 'Fob not recognized. Scan again.');
  const token = await issueToken(env, randomToken(), 'fob', { fob_id: fob });
  const claims = await verifyToken(env, token, 'fob');
  const url = `${origin(env)}/keyfob/bind?token=${token}`;
  const qr = qrcode(0, 'M');
  qr.addData(url);
  qr.make();
  await cleanupFobClaims(env);
  const saved = await env.DB.prepare(`INSERT INTO fob_claims(id, fob_id, created, expires)
    SELECT ?, ?, ?, ? WHERE (SELECT count(*) FROM fob_claims WHERE created > ?) < 30 RETURNING id`)
    .bind(claims.sub, fob, claims.iat, claims.exp, now() - 60).first();
  if (!saved) throw new HttpError(429, 'Too many scans. Please wait a minute.');
  return json({ token, url, expires: claims.exp, qr: `data:image/svg+xml;base64,${btoa(qr.createSvgTag({ cellSize: 6, margin: 24 }))}` }, 201);
}

export async function bindFob(request, env) {
  const params = new URL(request.url).searchParams;
  if (params.getAll('token').length !== 1) throw new HttpError(400, 'Scan a fob enrollment QR code at the makerspace first.');
  const token = params.get('token');
  const claim = await fobClaim(env, token);
  if (claim.claimed_by !== null) throw new HttpError(409, 'This QR code has already been used. Scan your fob again.');
  const member = await signedInMember(request, env);
  if (!member) {
    if (request.method === 'GET') return startLogin(request, env, 'member');
    throw new HttpError(401, 'Your sign-in expired. Reload this enrollment link to sign in again.');
  }
  const csrf = await hash(`fob-csrf:${cookie(request, 'thelab_member')}:${token}`);
  if (request.method === 'GET') {
    return page('Link your fob', `<p>Signed in as <strong>${e(member.discord_username)}</strong> (Discord ${e(member.discord_user_id)}).</p><p>Link fob <strong>${claim.fob_id}</strong> to your membership.</p>
      ${member.fob_id && member.fob_id !== claim.fob_id ? `<p>This replaces fob <strong>${member.fob_id}</strong>. Your previous fob will lose your membership’s door access after synchronization.</p>` : ''}
      <p>This code expires at ${e(new Date(claim.expires * 1000).toISOString())}. Door access follows your membership and waiver status.</p>
      <form method="post" action="/keyfob/bind?token=${e(token)}"><input type="hidden" name="csrf" value="${csrf}"><button class="btn btn-primary" type="submit">Link fob</button></form>`);
  }
  sameOrigin(request, env);
  if (request.headers.get('Content-Type')?.split(';')[0] !== 'application/x-www-form-urlencoded') throw new HttpError(415, 'Submit the enrollment form.');
  const form = new URLSearchParams(await boundedText(request, 2048));
  if (form.getAll('csrf').length !== 1 || form.get('csrf') !== csrf) throw new HttpError(403, 'Invalid enrollment form. Reload this page and try again.');
  await coordinated(env, member.member_id, 'linkFob', { token, discord_user_id: member.discord_user_id, auth_version: member.auth_version });
  return page('Fob linked', `<p role="status">Fob ${claim.fob_id} is now linked to your membership.</p><p>Door access follows your membership and waiver status. Door changes take effect after synchronization.</p><p>You can close this page.</p>`);
}

File: edgeproxy/kiosk.go
package main

import (
	"crypto/rand"
	"database/sql"
	_ "embed"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	qrcode "github.com/skip2/go-qrcode"
)

//go:embed kiosk.html
var kioskHTML string

//go:embed kiosk.js
var kioskJS string

//go:embed kiosk.css
var kioskCSS string

//go:embed kiosk-favicon.svg
var kioskIcon string

func kioskHeaders(w http.ResponseWriter) {
	w.Header().Set("Cache-Control", "no-store")
	w.Header().Set("Referrer-Policy", "no-referrer")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	w.Header().Set("X-Robots-Tag", "noindex, nofollow, noarchive")
	w.Header().Set("Content-Security-Policy", "default-src 'none'; script-src 'self'; style-src 'self'; img-src 'self' data:; connect-src 'self'; frame-ancestors 'none'; base-uri 'none'; form-action 'none'")
}

func kioskAsset(w http.ResponseWriter, r *http.Request) {
	kioskHeaders(w)
	asset := map[string]struct{ contentType, body string }{
		"/kiosk":             {"text/html; charset=utf-8", kioskHTML},
		"/kiosk/app.js":      {"text/javascript; charset=utf-8", kioskJS},
		"/kiosk/style.css":   {"text/css; charset=utf-8", kioskCSS},
		"/kiosk/favicon.svg": {"image/svg+xml", kioskIcon},
	}[r.URL.Path]
	w.Header().Set("Content-Type", asset.contentType)
	_, _ = io.WriteString(w, asset.body)
}

type kioskClaim struct {
	ID      string `json:"id"`
	Fob     uint32 `json:"fob_id"`
	Created int64  `json:"created"`
	Expires int64  `json:"expires"`
}

func kioskJSON(w http.ResponseWriter, status int, value any) {
	kioskHeaders(w)
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(value)
}

func kioskError(w http.ResponseWriter, status int, message string) {
	kioskJSON(w, status, map[string]string{"error": message})
}

func (e *edge) readKioskClaim(w http.ResponseWriter, r *http.Request) *kioskClaim {
	values := r.URL.Query()["token"]
	if len(values) != 1 || len(values[0]) != 64 || strings.Trim(values[0], "0123456789abcdef") != "" {
		kioskError(w, http.StatusGone, "This code is invalid or expired. Scan your fob again.")
		return nil
	}
	var claim kioskClaim
	err := e.db.QueryRowContext(r.Context(), "SELECT id, fob, created, expires FROM kiosk_claims WHERE id = ? AND expires > ?", values[0], time.Now().Unix()).Scan(&claim.ID, &claim.Fob, &claim.Created, &claim.Expires)
	if errors.Is(err, sql.ErrNoRows) {
		kioskError(w, http.StatusGone, "This code expired. Scan your fob again.")
		return nil
	}
	if err != nil {
		kioskError(w, http.StatusServiceUnavailable, "Fob enrollment is temporarily unavailable. Please scan again.")
		return nil
	}
	return &claim
}

// Only the authenticated tunnel listener exposes the fob behind a token.
func (e *edge) getKioskClaim(w http.ResponseWriter, r *http.Request) {
	if claim := e.readKioskClaim(w, r); claim != nil {
		kioskJSON(w, http.StatusOK, claim)
	}
}

func (e *edge) issueKioskClaim(w http.ResponseWriter, r *http.Request) {
	origin, err := url.Parse(r.Header.Get("Origin"))
	if err != nil || origin.Host != r.Host || (origin.Scheme != "http" && origin.Scheme != "https") || origin.User != nil || origin.Path != "" || origin.RawQuery != "" || origin.Fragment != "" {
		kioskError(w, http.StatusForbidden, "Reload the kiosk and try again.")
		return
	}
	if strings.Split(r.Header.Get("Content-Type"), ";")[0] != "application/json" {
		kioskError(w, http.StatusUnsupportedMediaType, "Fob not recognized. Scan again.")
		return
	}
	var input struct {
		Fob uint32 `json:"fob_id"`
	}
	body, err := io.ReadAll(http.MaxBytesReader(w, r.Body, 1024))
	if err != nil || decodeJSON(body, &input) != nil || input.Fob == 0 {
		kioskError(w, http.StatusBadRequest, "Fob not recognized. Scan again.")
		return
	}
	if e.workerAuth == nil {
		kioskError(w, http.StatusServiceUnavailable, "Fob enrollment is not configured. Please ask a member for help.")
		return
	}
	var nonce [32]byte
	_, _ = rand.Read(nonce[:])
	created := time.Now().Unix()
	claim := kioskClaim{hex.EncodeToString(nonce[:]), input.Fob, created, created + 300}
	link := e.workerAuth.issuer + "/keyfob/bind?token=" + claim.ID
	qr, err := qrcode.Encode(link, qrcode.Medium, 512)
	if err == nil {
		err = e.transaction(r.Context(), func(tx *sql.Tx) error {
			if _, err := tx.Exec("DELETE FROM kiosk_claims WHERE expires <= ?", claim.Created); err != nil {
				return err
			}
			return tx.QueryRow(`INSERT INTO kiosk_claims(id, fob, created, expires)
				SELECT ?, ?, ?, ? WHERE (SELECT count(*) FROM kiosk_claims WHERE created > ?) < 30 RETURNING id`,
				claim.ID, claim.Fob, claim.Created, claim.Expires, claim.Created-60).Scan(&claim.ID)
		})
	}
	if errors.Is(err, sql.ErrNoRows) {
		kioskError(w, http.StatusTooManyRequests, "Too many scans. Please wait a minute.")
		return
	}
	if err != nil {
		kioskError(w, http.StatusServiceUnavailable, "Fob enrollment is temporarily unavailable. Please scan again.")
		return
	}
	kioskJSON(w, http.StatusCreated, map[string]any{"token": claim.ID, "url": link, "expires": claim.Expires, "qr": "data:image/png;base64," + base64.StdEncoding.EncodeToString(qr)})
}

func (e *edge) kioskClaimStatus(w http.ResponseWriter, r *http.Request) {
	claim := e.readKioskClaim(w, r)
	if claim == nil {
		return
	}
	if e.workerAuth != nil {
		req, err := http.NewRequestWithContext(r.Context(), "GET", e.workerAuth.issuer+"/keyfob/status?token="+claim.ID, nil)
		if err == nil {
			response, err := e.workerAuth.client.Do(req)
			if err == nil {
				defer response.Body.Close()
				body, err := io.ReadAll(io.LimitReader(response.Body, 1025))
				var result struct {
					Claimed *bool `json:"claimed"`
				}
				if err == nil && len(body) <= 1024 && response.StatusCode == http.StatusOK && decodeJSON(body, &result) == nil && result.Claimed != nil {
					kioskJSON(w, http.StatusOK, map[string]any{"claimed": *result.Claimed, "expires": claim.Expires})
					return
				}
			}
		}
	}
	kioskError(w, http.StatusServiceUnavailable, "Could not check completion. Retrying…")
}
